text/0000-float-semantics.md

```diff
@@ -0,0 +1,260 @@
+
+# Motivation
+[motivation]: #motivation
+
+We have a plethora of open issues that boil down to "is this sequence of float operations allowed to produce the given result".
+This is caused by a combination of surprising effects introduced by LLVM optimizations, bugs in MIR and LLVM optimizations, and bugs (or at least non-conformance) in certain targets. See [here](https://github.com/rust-lang/unsafe-code-guidelines/issues/237) for a collection of issues.
+
+It's time to stop leaving our users in the dark about what actually is and is not guaranteed.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Primitive operations on floating-point types generally produce results that exactly match IEEE 754-2008:
+if you never use `to_bits`, `copysign`, `is_sign_negative`, or `is_sign_positive` on a NaN, and don't construct a NaN using `from_bits`, nor use any unsafe operations that are equivalent to these safe methods, then your code will behave perfectly deterministically and according to the IEEE specification.
+
+If you *do* use these operations on NaNs, then the exact behavior you see can depend on compiler version, compiler flags, target architecture, and it can even be non-deterministic (i.e., running the same operation on the same inputs twice can yield different results).
+The results produced in these cases do *not* always conform to the IEEE specification.
+See the reference section for what exactly is guaranteed.
+
+When a floating-point value is just passed around, its contents (including the bits of a NaN) do *not* change.
+
+When you use a floating-point operation in [`const` context](https://doc.rust-lang.org/reference/const_eval.html#const-context), the same specification applies: NaN bit patterns are non-deterministic.
+In particular, the bit pattern produced at compile-time can differ from the bit pattern produced by the same operation at run-time.
+
+Certain targets unfortunately are known to not implement these semantics precisely (see [below](#target-specific-problems)).
+The [platform support page](https://doc.rust-lang.org/rustc/platform-support.html) will list those caveats.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC specifies the behavior of `+`, `-` (unary and binary), `*`, `/`, `%`, `abs`, `copysign`, `mul_add`, `sqrt`, `as`-casts that involve floating-point types, and all comparison operations on floating-point types.
+Here, "floating-point types" are `f32` and `f64` and all similar types that might be added in the future such as `f16`, `f128`.
+Except for the cases handled below, these operations produce results that exactly match IEEE 754-2008 (with roundTiesToEven and default exception handling, without abruptUnderflow/flush-to-zero).
```

Comment:
I'm actually a bit confused about this point, I copied from somewhere that roundTiesToEven is the default, but `1.8 as i32` returns `1` so actually the default seems to be toward-zero?

Reply:
no, all operations other than conversion to integer default to round-to-nearest-ties-to-even, conversion to integer defaults to truncate (you can think of it as specifically overriding rounding mode for conversion to integer). This is inherited from C afaict, but is standard practice across a wide range of languages. the default for conversion to integer can be overridden using `floor`/`ceil`/`round`/`trunc`/`round_even` (`x.floor() as i32` rounds toward -Infinity)

Reply:
Oh okay, so this affects *all* operations? But the hardware instructions that we use for float-to-int casts ignore the rounding mode and always round to zero?

Reply:
It affects almost all operations, some use specified rounding modes (e.g. `floor` uses round toward -Infinity).
LLVM intentionally picks round-to-zero instructions for float-to-int casts. e.g. on `x86_64-unknown-linux-gnu`, LLVM picks `cvttss2si` (which rounds to zero) instead of `cvtss2si` (which uses the dynamic rounding mode); and on `i586-unknown-linux-gnu`, LLVM temporarily changes the rounding mode to round toward zero while doing the conversion.

Reply:
Makes sense, thanks!